### PR TITLE
Feature/add ve perp reward distributor

### PR DESCRIPTION
**Pull request by 42bchen**

### CODE REVIEW

**Review by Detoo — APPROVED** [review]

File: contracts/vePERPRewardDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    /// @param liquidityProvider Liquidity provider address
+    /// @param week Week number of the reward claimed
+    /// @param claimedBalance Amount of vePERP reward claimed
+    /// @param merkleProof Merkle proof of the week's allocation
+    function claimWeek(
+        address liquidityProvider,
+        uint256 week,
+        uint256 claimedBalance,
+        bytes32[] calldata merkleProof
+    ) public override userLockTimeCheck(liquidityProvider) {
```
Comment (tailingchen) [inline contracts/vePERPRewardDistributor.sol:123]: Internal audit:
Why do we need the modifier userLockTimeCheck?
I saw there is [lock-time checking](https://github.com/perpetual-protocol/voting-escrow/blob/main/contracts/vePERP.vy#L473) in `deposit_for` as well. What't the difference between these two?

[the remark this answers is not in the document]
Comment (42bchen) [reply contracts/vePERPRewardDistributor.sol:123]: reward will store in this contract, and if user want to claim the reward, he needs to create a lock with lock time > `_minLockDuration`

File: contracts/vePERPRewardDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        token.approve(vePERPArg, uint256(-1));
+    }
+
+    function seedAllocations(
+        uint256 week,
+        bytes32 merkleRoot,
+        uint256 totalAllocation
+    ) public override onlyOwner {
+        // vePRD_TIZ: total allocation is zero
+        require(totalAllocation > 0, "vePRD_TIZ");
```
[the remark this answers is not in the document]
Comment (42bchen) [reply contracts/vePERPRewardDistributor.sol:88]: not sure if `totalAllocation` is zero, we still need to do the `seedAllocation` or not, the contract support `seedAllocation` on nonsequential weeks, so if `totalAllocation` is zero, I think we can skip `seedAllocations` on that week, and we can check week status via `merkleRootIndexes`

[the remark this answers is not in the document]
Comment (tailingchen) [reply contracts/vePERPRewardDistributor.sol:88]: My opinion is seed allocation for every week should be uploaded on chain even if the total reward is zero. It's kind of proof.

**Pull request merged**